app/actions/index.js

```diff
@@ -33,6 +33,119 @@ export const createDat = () => dispatch => {
   addDat({ path })(dispatch)
 }
 
+export const showDownloadScreen = key => ({
+  type: 'SHOW_DOWNLOAD_SCREEN',
+  key: encode(key)
+})
+export const hideDownloadScreen = () => ({ type: 'HIDE_DOWNLOAD_SCREEN' })
+export const cancelDownloadDat = key => dispatch => {
+  key = encode(key)
+  const { dat } = dats[key]
+
+  for (const con of dat.network.connections) {
+    con.removeAllListeners()
+  }
+  dat.stats.removeAllListeners()
+  clearInterval(dat.updateInterval)
+
+  dat.close()
+  delete dats[key]
+  dispatch({ type: 'REMOVE_DAT', key })
+}
+
+export const changeDownloadPath = key => dispatch => {
+  const files = remote.dialog.showOpenDialog({
+    properties: ['openDirectory']
+  })
+  if (!files || !files.length) return
+  const path = files[0]
+  dispatch({ type: 'CHANGE_DOWNLOAD_PATH', key, path })
+}
+
+export const downloadSparseDat = ({ key }) => dispatch => {
+  if (key) key = encode(key)
+  const path = `${homedir()}/Downloads/${key}`
+  const paused = false
+
+  if (key) dispatch({ type: 'ADD_DAT', key, path, paused })
+  const opts = {
+    watch: true,
+    resume: true,
+    ignoreHidden: true,
+    compareFileContent: true
+  }
+
+  Dat(path, { key, sparse: true }, (error, dat) => {
+    if (error) return dispatch({ type: 'ADD_DAT_ERROR', key, error })
+    if (!key) {
+      key = encode(dat.key)
+      dispatch({ type: 'ADD_DAT', key, path, paused })
+    }
+
+    dat.trackStats()
+
+    dispatch({
+      type: 'DAT_METADATA',
+      key,
+      metadata: {
+        title: basename(path),
+        author: 'Anonymous'
+      }
+    })
+
+    dispatch({ type: 'ADD_DAT_SUCCESS', key })
+
+    dat.archive.readFile('/dat.json', (err, blob) => {
+      if (err) return
+
+      let metadata = {}
+      try {
+        metadata = JSON.parse(blob)
+      } catch (_) {}
+
+      dispatch({ type: 'DAT_METADATA', key, metadata })
+    })
+
+    const walk = () => {
```

Comment:
This seems _(unnecessarily)_ copy & pasted from addDat. Please extract into common code.

Reply:
+1